File: apps/api/src/platform/encryption.ts
import { createCipheriv, createDecipheriv, randomBytes } from "node:crypto";

/**
 * AES-256-GCM constants.
 * GCM provides both confidentiality (AES-256) and integrity (auth tag).
 */
const ALGO = "aes-256-gcm" as const;
const IV_BYTES = 12; // 96-bit IV — optimal for GCM
const TAG_BYTES = 16; // 128-bit auth tag

/**
 * Parses the base64-encoded 32-byte AES-256 encryption key from the environment.
 * Throws if the decoded key is not exactly 32 bytes.
 * @param base64 - Base64-encoded key string (TOKEN_ENC_KEY in .env).
 * @returns A 32-byte {@link Buffer}.
 * @throws If the key is empty or not 32 bytes after decoding.
 */
export function parseEncryptionKey(base64: string): Buffer {
  if (!base64) {
    throw new Error("TOKEN_ENC_KEY is not set; cannot encrypt or decrypt repo tokens.");
  }
  const key = Buffer.from(base64, "base64");
  if (key.byteLength !== 32) {
    throw new Error(
      `TOKEN_ENC_KEY must decode to exactly 32 bytes for AES-256 (got ${key.byteLength}).`,
    );
  }
  return key;
}

/**
 * Encrypts a plaintext string using AES-256-GCM.
 * Produces a compact hex string in the format: `<iv>:<ciphertext>:<authTag>`.
 * @param plaintext - The value to encrypt (e.g. a repo deploy token).
 * @param key - 32-byte AES key produced by {@link parseEncryptionKey}.
 * @returns Hex-encoded `iv:ciphertext:authTag` suitable for storage in Postgres.
 */
export function encryptToken(plaintext: string, key: Buffer): string {
  const iv = randomBytes(IV_BYTES);
  const cipher = createCipheriv(ALGO, key, iv);
  const encrypted = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()]);
  const tag = cipher.getAuthTag();
  return `${iv.toString("hex")}:${encrypted.toString("hex")}:${tag.toString("hex")}`;
}

/**
 * Decrypts an AES-256-GCM ciphertext produced by {@link encryptToken}.
 * @param ciphertext - Hex-encoded `iv:ciphertext:authTag` string.
 * @param key - 32-byte AES key produced by {@link parseEncryptionKey}.
 * @returns The original plaintext.
 * @throws If the format is invalid or the GCM auth tag verification fails.
 */
export function decryptToken(ciphertext: string, key: Buffer): string {
  const parts = ciphertext.split(":");
  if (parts.length !== 3) {
    throw new Error("Invalid ciphertext format; expected iv:ciphertext:authTag.");
  }
  const [ivHex, encHex, tagHex] = parts as [string, string, string];
  const iv = Buffer.from(ivHex, "hex");
  const encrypted = Buffer.from(encHex, "hex");
  const tag = Buffer.from(tagHex, "hex");
  if (tag.byteLength !== TAG_BYTES) {
    throw new Error(`Invalid auth tag length: ${tag.byteLength} (expected ${TAG_BYTES}).`);
  }
  const decipher = createDecipheriv(ALGO, key, iv);
  decipher.setAuthTag(tag);
  return decipher.update(encrypted) + decipher.final("utf8");
}

File: apps/rag/src/services/encryption.py
"""AES-256-GCM repo token decryption (matches apps/api platform/encryption.ts)."""

from __future__ import annotations

import base64

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

IV_BYTES = 12
TAG_BYTES = 16


def parse_encryption_key(base64_key: str) -> bytes:
    """Decode the base64 TOKEN_ENC_KEY into a 32-byte AES key.

    @param base64_key - Base64-encoded key from the environment.
    @returns 32-byte key material.
    @raises ValueError when the key is missing or not 32 bytes after decoding.
    """
    if not base64_key:
        raise ValueError("TOKEN_ENC_KEY is not set; cannot decrypt repo tokens.")
    key = base64.b64decode(base64_key)
    if len(key) != 32:
        raise ValueError(f"TOKEN_ENC_KEY must decode to 32 bytes (got {len(key)}).")
    return key


def decrypt_token(ciphertext: str, key: bytes) -> str:
    """Decrypt an `iv:ciphertext:authTag` hex string produced by the Node API.

    @param ciphertext - Encrypted token string stored in `repos.token_enc` (UTF-8 bytes).
    @param key - 32-byte AES key from {@link parse_encryption_key}.
    @returns Plaintext deploy token.
    @raises ValueError when the format or auth tag is invalid.
    """
    parts = ciphertext.split(":")
    if len(parts) != 3:
        raise ValueError("Invalid ciphertext format; expected iv:ciphertext:authTag.")
    iv_hex, enc_hex, tag_hex = parts
    iv = bytes.fromhex(iv_hex)
    encrypted = bytes.fromhex(enc_hex)
    tag = bytes.fromhex(tag_hex)
    if len(tag) != TAG_BYTES:
        raise ValueError(f"Invalid auth tag length: {len(tag)} (expected {TAG_BYTES}).")
    aesgcm = AESGCM(key)
    plaintext = aesgcm.decrypt(iv, encrypted + tag, None)
    return plaintext.decode("utf-8")


def token_bytes_to_ciphertext(token_enc: bytes | None) -> str | None:
    """Convert BYTEA column bytes to the hex ciphertext string used by decrypt_token.

    @param token_enc - Raw bytes from Postgres or `None`.
    @returns UTF-8 decoded ciphertext string or `None`.
    """
    if token_enc is None:
        return None
    return token_enc.decode("utf-8")
